Introduce multiple versions of ZLS

jmbaur opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

ZLS versions prefer to be kept in sync with the Zig version in use. By introducing multiple ZLS versions, we can keep multiple ZLS variants in the package-set, similar to what is done for Zig. In addition, we can add an alias for whichever ZLS version matches the current zig alias. 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jmbaur commented:
related to #436560

Guanran928 commented on pkgs/top-level/aliases.nix:
```diff
@@ -2554,6 +2554,7 @@ mapAliases {
   zimlib = throw "'zimlib' has been removed because it was an outdated and unused version of 'libzim'"; # Added 2025-03-07
   zinc = zincsearch; # Added 2023-05-28
   zint = zint-qt; # Added 2025-05-15
+  zls = zls_0_14; # Added 2025-08-31
```
What if people enabled `allowAliases = false`? I think this should be done in `all-packages.nix` instead.

jmbaur commented:
@Guanran928 any other thoughts on this one?

Ratakor commented:
Fixes #392896

jmbaur committed:
zls_0_14: init at 0.14.0

This moves zls out of by-name in preparation for adding more zls
versions with shared nix expressions.

jmbaur committed:
zls_0_15: init at 0.15.0

jmbaur committed:
zls: add jmbaur as maintainer

kirillrdy commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9262,6 +9262,13 @@ with pkgs;
   zigStdenv = if stdenv.cc.isZig then stdenv else lowPrio zig.passthru.stdenv;
 
+  inherit (callPackages ../development/tools/zls { })
+    zls_0_14
```
shouldn't we also expose `zls_0_15` ?

jmbaur replied:
ah of course, my mistake! fixed

kirillrdy reviewed (approved):
i am pretty sure nixpkgs review is broken

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 438854`
Commit: `e6ff273e692674c3e9e078f1f1200f2eca4df63a`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>chatgpt-retrieval-plugin</li>
    <li>chatgpt-retrieval-plugin.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 35 packages built:</summary>
  <ul>
    <li>fiddler-everywhere</li>
    <li>flare-signal</li>
    <li>gitu</li>
    <li>lazygit</li>
    <li>litestar (python313Packages.litestar)</li>
    <li>litestar.dist (python313Packages.litestar.dist)</li>
    <li>lmstudio</li>
    <li>perfect_dark</li>
    <li>python312Packages.fast-query-parsers</li>
    <li>python312Packages.fast-query-parsers.dist</li>
    <li>python312Packages.litestar</li>
    <li>python312Packages.litestar.dist</li>
    <li>python312Packages.piccolo</li>
    <li>python312Packages.piccolo.dist</li>
    <li>python312Packages.targ</li>
    <li>python312Packages.targ.dist</li>
    <li>python312Packages.weaviate-client</li>
    <li>python312Packages.weaviate-client.dist</li>
    <li>python313Packages.fast-query-parsers</li>
    <li>python313Packages.fast-query-parsers.dist</li>
    <li>python313Packages.piccolo</li>
    <li>python313Packages.piccolo.dist</li>
    <li>python313Packages.targ</li>
    <li>python313Packages.targ.dist</li>
    <li>python313Packages.weaviate-client</li>
    <li>python313Packages.weaviate-client.dist</li>
    <li>qucs-s</li>
    <li>redpanda-client</li>
    <li>rime-wanxiang</li>
    <li>sparkle</li>
    <li>taze</li>
    <li>vscode-extensions.rooveterinaryinc.roo-cline</li>
    <li>wapiti</li>
    <li>wapiti.dist</li>
    <li>zls_0_15</li>
  </ul>
</details>
